src/com/ms/silverking/cloud/dht/client/impl/AsyncOperationImpl.java

```diff
@@ -27,446 +28,490 @@
+      opState = getState();
+      if (_listenStates.contains(opState)) {
+        notifyListeners = true;
+      }
+      // We must add this to listeners if either:
+      //   1 - We are not notifying listeners now
+      //   2 - This call requests notification of incomplete state
+      //       (This will trigger both immediate callback and subsequent callbacks)
+      if (!notifyListeners || _listenStates.contains(OperationState.INCOMPLETE)) {
+        if (listeners == null) {
+          listeners = new HashSet<>();
         }
-    }    
-    
-    void _waitForCompletion() throws OperationException {
-        lock.lock();
-        try {
-            while (!result.isComplete()) {
-                try {
-                    Log.fine("activeOp awaiting ", this);
-                    if (!spin) {
-                        cv.await();
-                    } else {
-                        cv.awaitNanos(spinDurationNanos);
-                    }
-                } catch (InterruptedException ie) {
-                }
-            }
-            if (isFailure(result)) {
-                throwFailedException();
-            }
-        } finally {
-            lock.unlock();
+        for (AsyncOperationListener listener : _listeners) {
+          listeners.add(new Pair<>(listener, _listenStates));
         }
+        opsWithListeners.put(getUUID(), this);
+      }
+    } finally {
+      lock.unlock();
     }
-    
-    public boolean poll() {
-        boolean    complete;
-        
-        lock.lock();
-        try {
-            complete = result.isComplete();
-        } finally {
-            lock.unlock();
-        }
-        return complete;
+    if (notifyListeners) {
+      // Using calling thread to trigger updates for completion, but only for completion
+      if (opState != OperationState.INCOMPLETE) {
+        notifyListeners(_listeners);
+      } else {
+        notificationWorker.update(this, ImmutableSet.copyOf(_listeners));
+      }
     }
-    
-    public boolean poll(long duration, TimeUnit timeUnit) {
-        boolean    complete;
-        
-        lock.lock();
+  }
+
+  void _waitForCompletion() throws OperationException {
+    lock.lock();
+    try {
+      while (!result.isComplete()) {
         try {
-            while (!result.isComplete()) {
-                try {
-                    if (!spin) {
-                        cv.await(duration, timeUnit);
-                    } else {
-                        cv.awaitNanos(spinDurationNanos);
-                    }
-                } catch (InterruptedException ie) {
-                }
-            }
-            complete = result.isComplete();
-        } finally {
-            lock.unlock();
+          Log.fine("activeOp awaiting ", this);
+          if (!spin) {
+            cv.await();
+          } else {
+            cv.awaitNanos(spinDurationNanos);
+          }
+        } catch (InterruptedException ie) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)